lb_test.go

```diff
@@ -583,3 +587,178 @@ run:
+	// Prepare TLS data for the test
+	err := tlsutil.CreateLocalTLSData("test")
+	defer func() {
+		t.Log("test unwind, delete TLS files")
+		out, err := tlsutil.WipeLocalTLSData("./")
+		if err != nil {
+			t.Error("cannot clean pre-arranged test files")
+		}
+		t.Logf("files deleted: %+v", out)
+	}()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Prepare responding servers for the test
+	// TODO: Optimize tests in the way to obtain random ranges for ports to configure LB and test targets
+	t.Log("test prepare, create responding servers")
+	stopServer1, err := httputil.CreateTestServer(9095, "api", "Server 1 responded")
+	if err != nil {
+		t.Errorf("Failed to start test server 1: %v", err)
+	}
+	defer stopServer1()
+	stopServer2, err := httputil.CreateTestServer(9096, "api", "Server 2 responded")
+	if err != nil {
+		t.Errorf("Failed to start test server 2: %v", err)
+	}
+	defer stopServer2()
+	stopServer3, err := httputil.CreateTestServer(9097, "api", "Server 3 responded")
+	if err != nil {
+		t.Errorf("Failed to start test server 3: %v", err)
+	}
+	defer stopServer3()
+
+	t.Log("test prepare, create loadbalancer instance")
+	cert, err := os.ReadFile("server.crt")
+	if err != nil {
+		t.Fatal(err)
+		return
+	}
+	key, err := os.ReadFile("server.key")
+	if err != nil {
+		t.Fatal(err)
+		return
+	}
+	ca, err := os.ReadFile("ca.crt")
+	if err != nil {
+		t.Fatal(err)
+		return
+	}
+
+	const quotaPerSecond = 10
+
+	balancer, err := NewLoadBalancer(ctx, []ServicePool{{
+		SvcIdentity:          "test",
+		SvcPort:              9098,
+		SvcRateQuotaTimes:    quotaPerSecond,
+		SvcRateQuotaDuration: time.Second * 1,
+		SvcRoutes: []ServicePoolRoute{
+			ServicePoolRoute{
+				ServicePath:   "localhost:9095",
+				ServiceActive: true,
+			},
+			ServicePoolRoute{
+				ServicePath:   "localhost:9096",
+				ServiceActive: true,
+			},
+			ServicePoolRoute{
+				ServicePath:   "localhost:9097",
+				ServiceActive: true,
+			},
+		},
+		Certificate: string(cert),
+		CertKey:     string(key),
+		CACert:      string(ca),
+	}}, Options{})
+	if err != nil {
+		t.Fatal("cannot configure load balancer")
+	}
+
+	// Give 10 seconds for testing everything
+	go func() {
+		<-time.After(time.Second * 15)
+		cancelAll()
+	}()
+
+	// Create senders
+	responses := make(chan string, 200)
+
+	// Create tools to calculate tick counts
+	// TODO: For proper calculation in the test we need to collect all responses by time and then take ranges as following
+	//   for each object like: {response, time}
+	//      from the start - collect range until it fits quota (like all items within 1s)
+	//           collect amount of items in such range to verify closest number of requests approved by quota
+	lastResponseCount := 0
+	tickCounts := make([]int, 0)
+
+	// Currently test strategy is to approximately calculate that there were +/- 10 requests per second
+	go func() {
+
+		// Give 5 seconds for everything to startup
+		<-time.After(time.Second * 5)
```

Comment:
I'm cautious of tests which rely on fixed sleeps like these. You typically end in in one of two scenarios:

- Your estimated time is too long, and the test takes longer to run than it needs to. Over thousands of tests this can add up to significant additional waiting time for engineers.
- Your estimated time is too short/just right, but when running on a busy CI system, the test begins to flake.

I prefer tests which make use of "Eventually"-esque loops, these can try a condition every 100ms or so, and then allow the test to proceed once the condition is met. A fallback max time limit can be set, and this can be more generous than the wait time you would have used before.